src/circuit/circuit_frame.rs

```diff
@@ -3075,6 +3439,108 @@ fn apply_continuation<F: LurkField, CS: ConstraintSystem<F>>(
     Ok((result_expr, result_env, result_cont, make_thunk_num))
 }
 
+fn hide<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    secret: F,
+    maybe_payload: &AllocatedPtr<F>,
+    store: &Store<F>,
+) -> Result<AllocatedPtr<F>, SynthesisError> {
+    let hide_ptr = if let Some(ptr) = maybe_payload.ptr(store).as_ref() {
+        match store.hidden(secret, *ptr) {
+            Some(c) => c,
+            None => store.get_nil(),
+        }
+    } else {
+        store.get_nil()
+    };
+    AllocatedPtr::alloc_ptr(&mut cs.namespace(|| "hide"), store, || Ok(&hide_ptr))
+}
+
+fn open<F: LurkField, CS: ConstraintSystem<F>>(
+    mut cs: CS,
+    maybe_commit: &AllocatedPtr<F>,
+    store: &Store<F>,
+) -> Result<AllocatedPtr<F>, SynthesisError> {
+    let hash = match maybe_commit.hash().get_value() {
+        Some(h) => h,
+        None => F::zero(),
+    };
+    let open_ptr = match store.get_maybe_opaque(Tag::Comm, hash) {
+        Some(c) => match store.open(c) {
+            Some(o) => o,
+            None => store.get_nil(),
+        },
+        None => store.get_nil(),
```

Comment:
See comment above in definition of `hide`.

Reply:
Actually, this is yet another case — and I believe here it is actually incorrect to always return NIL. In some cases, it is incorrect to return *any* value. In those cases, we should either panic or (maybe?) return `Err::SynthesisError::Unsatisfiable` (though I think that is not quite right).

The important point to consider is that if we try to evaluate an opening for which no preimage is known (for example `(open 123)`) then it should not be possible to generate *any* proof. We cannot create a proof that this yields an error because there may well be an opening the prover doesn't know of. But we also cannot create a proof which returns any non-error value because the prover doesn't know it.

If you want to see this bug in action, you can modify `eval.rs` to create the necessary inputs. (In its usual form, Lurk evaluation will never create evaluation frames/witnesses for such an opening because it is invalid.)

Change line 974:
```
                    Op1::Open => store.open_mut(*result).unwrap_or(store.nil()),
```

Then add this test to `nova.rs`:
```
    #[test]
    fn outer_prove_bad_open() {
        let s = &mut Store::<Fr>::default();
        let expr = "(open 123))";
        let expected = s.nil();
        let terminal = s.get_cont_terminal();
        nova_test_aux(s, expr, Some(expected), None, Some(terminal), None, 2);
    }
```

Obviously the above should *never* yield a terminal continuation. Yet the test will pass.

To see clearly that this bug relies on the wrong value being returned in the code here, alter the modification in `eval.rs` to return `store.get_t()` instead of `store.nil()`. Now the same test will fail because of a return-value mismatch.

---

Of course we *also* have to deal with the actually dummy case in which the code path is not even being exercised in the circuit. I think it's interesting to note that the bug I've described just prior is exactly the kind of thing I enumerated in an earlier comment as something we need to guard against when conflating the different kinds of apparently 'dummy' values (and the reasons for them). It's a good and helpfully immediate example.

I think this highlights that we need to as clearly as possible distinguish between at least:
- constraints corresponding to unexercised code paths
- values corresponding to those in 'blank circuits'
- hash values whose preimages may or may not be known
- default values

We may need to try to be more formal about these distinctions — at least at the level of documenting with naming and comments; and wherever possible at the code level.

Reply:
For me it makes sense if `(open 123)` returns an error such as "trying open unknown commitment", because it represents the fact this opening is not in the store (and not that such an opening doesn't exist in general).

Reply:
I cannot agree with this perspective. It's an important property of the system that it be impossible to obtain valid proofs for different evaluation results of the same expression when using the 'deterministic' subset of the language. Currently the whole language is deterministic in the way I mean. That doesn't mean that non-deterministic constructs aren't used under the hood (that's how all the hash-consing works, obviously) — but since we rely on the hashes used being computationally binding we consider that there exists at-most one preimage available to a computationally-bound prover. If the prover does not know that preimage, he can't create a proof.

If this weren't the case, anyone could create proofs over other people's private data — showing that the computations yield errors. There is a real case in which such a proof would be useful: to demonstrate that some operation over private data *does* indeed lead to an error. That useful case would be impossible if things work as you describe. In fact, it would become easy to create proofs that evaluation of *many* valid expressions result in errors if we generalized the principal (of using absence or presence in the store — which we can consider a synonym for 'prover knowledge') as the determining factor in whether an expression evaluation should lead to an error.

You can see that we handle this correctly in the case of `car` and `cdr` operations: https://github.com/lurk-lang/lurk-rs/blob/master/src/store.rs#L1809-L1810 since we panic and will not produce a proof if attempting to destructure an unknown pair. This could be made more graceful and perhaps clearer by returning an explicit error rather than panicking. However, that would require reworking the circuit-synthesis code to deal with such errors gracefully. (We may want to consider doing this, but I think it's not trivial. For now panicking is 'safe' and correct.)

Reply:
If we panic instead of returning something, then we break many other unit tests, because it should behave as a dummy in those situations. However, even if I return a new bogus dummy pointer, the problem you mentioned would not be solve, because I can adapt your modifications above accordingly. For example, I can replace line 974 by 
```
Op1::Open => store.open_mut(*result).unwrap_or(store.bogus_dummy())
```
In other words, if someone can change `eval.rs`, then it is possible to make something similar to the unit test you suggested.